cpp/src/parquet/types.cc

```diff
@@ -259,6 +266,34 @@ std::string ConvertedTypeToString(ConvertedType::type t) {
   }
 }
 
+int32_t GetCompressionCodecDefaultCompressionLevel(Compression::type t) {
```

Comment:
Why is this necessary? Why not use the 2-argument form of `Codec::Create` is the compression level isn't specified?

Reply:
In PageWriter::Open I would like to have to the compression level already selected since there are two different paths selected:: BufferedPageWriter and SerializedPageWriter.
I do not want to introduce any default parameters and leave deduction of the compression level at a level close to the arrow::util::Codec::Create because this means there will be two similar paths maintained - one with the compression level selected and the one with no compression level.
Selecting the compression level as close to the parquet api as possible means that there will be for the most part a single path maintained. However, there is a complication that arrow maintains its own compression enum with an extra bz2 entry. Thus, there's a need for some translation like it was already done and a bit simplified in the newest iteration of the change.
I still followed your advice to some extent which also made the change slightly smaller.

Reply:
I think it makes more sense for each codec to "know" its own default compression level rather than localizing/externalizing this information into a function that deals with all the codecs.  From the parquet APIs, I think you could use placeholder value like -1 to signify using the default compression level.

Reply:
No, -1 cannot be a placeholder because it can be a valid compression level for some codec which can be different from the default in Arrow. That's why it's best to query the default compression level close to the API. We cannot have a magic value which signals to use the default value which is some other magic value.

Reply:
> That's why it's best to query the default compression level close to the API.

That doesn't follow. You could e.g. have both `ZSTDCodec()` and `ZSTDCodec(int compression_level)` constructors.

Reply:
A stack trace for creating ZSTDCodec which shows one possible path:
arrow::util::ZSTDCodec::ZSTDCodec
arrow::util::Codec::Create
parquet::GetCodecFromArrow
parquet::SerializedPageWriter::SerializedPageWriter
parquet::PageWriter::Open
parquet::RowGroupSerializer::NextColumn

How should be determined in arrow::util::Codec::Create which constructor to use? Once you answer that, please tell me what should happen parquet::GetCodecFromArrow. Then please do that for the next levels.
Is it somehow possible to select which constructor to use without duplicating paths?

A simpler solution is to determine the default compression level for the specified codec higher up in the stack, for example in parquet::RowGroupSerializer::NextColumn. Then we can make the compression_level explicit, as it is done in my patch, and there will be no need for duplicating code.

I can still add both constructors which will simplify a bit some of the code in a few tests which doesn't bring much benefit.

I would like to hear what you think of all of this.

Reply:
Another possibility is `ZSTDCodec(int compression_level = kDefaultCompressionLevel)` where `kDefaultCompressionLevel` is chosen such that it is unlikely to be confused with an actual compression level (for example `INT_MIN`).

Reply:
`ZSTDCodec(int compression_level = kDefaultCompressionLevel)` this is not any different from your early proposal of having two constructors. On higher levels in the stack it would still have to be determined whether to pass an argument or not.

I do not want to propagate some random value through the whole stack with the hope that it doesn't conflict with what the user might specify.

So, let's see. 20 users decide that they want to pass INT_MIN as a compression level. Sure, it doesn't make sense but it works! ZSTD would be set up to use a compression level of 1 since INT_MIN is this magic value which is a hint that arrow is free to select whatever it wants.
Then, arrow changes the implementation a bit so that instead of INT_MIN we pick the much nicer 0x31337 magic value.
All user code suddenly starts behaving in a different way.

Reply:
> On higher levels in the stack it would still have to be determined whether to pass an argument or not.

No, you could simply pass `kDefaultCompressionLevel`.

> So, let's see. 20 users decide that they want to pass INT_MIN as a compression level. Sure, it doesn't make sense but it works!

I am not very interested in discussing contrived scenarios.

Reply:
> No, you could simply pass kDefaultCompressionLevel.
Yes, but kDefaultCompressionLevel will be that magic value you suggest introducing.

> I am not very interested in discussing contrived scenarios.
It is a hack to have an internal representation which can conflict with what the user provides. Especially, if it can easily be avoided.

Reply:
> No, -1 cannot be a placeholder because it can be a valid compression level for some codec which can be different from the default in Arrow.

Ok, thanks for correcting me.  Another suggestion would be to add a method on `ColumnProperties` that can tell you whether to use the codec defaults or override it with a user-supplied value.  Something like:

`bool using_compression_level`

that only returns `true` when a user has manually set the compression level.  I've seen problems like this solved with `boost::optional` but feels like overkill for this case, but as you've seen I'm often wrong so maybe check that out too.